flake-module.nix

```
@@ -99,6 +99,19 @@ in
                     description = ''Package overrides given new source path'';
                     default = { };
                   };
+                  easy-overrides = mkOption {
+                    description = " An attribute set with haskell packages and the options.";
+                    type = with types; attrsOf (submodule {
```

```suggestion
                    type = with types; attrsOf (deferredModulesWith { staticModules = [ { old, ... }: {
```

While this is not a trivial change, it would allow the `old` parameter to be passed in as a module argument. The attributes in `old` could be used as part of the option default definition, bridging the gap between overrides and options.

---

IIUC, this would give us something like:

```nix
{
  easy-overrides = {
    aeson = old: {
      enableLibraryProfiling = ! old.enableLibraryProfiling;
    };
}
```

right? Thus solving @maralorn 's "Also wondering how to deal with overrides which need to access the old attributes." question below.

---

As long as the resulting API is not complex ...

---

@roberth Can you help me to understand to advantage of using `deferredModuleWith`, rather than something like below (which is working).

```nix
                  easy-overrides =
                    let settingsFormat = pkgs.formats.json { };
                    in
                    mkOption {
                      description = " An attribute set with haskell packages and the options.";
                      type = with types; submodule {
                        freeformType = oneOf [(attrsOf raw) (functionTo (attrsOf raw))];
                      };
                      default = { };
                    };
```

---

`submodule` returns the computed fixpoint `config`, whereas `deferredModule` lets the module be a module that's reusable, as a function.

Also `freeformType` doesn't really support anything other than an attrset at its root. It would never get to see the module function.

---

It's hard to communicate the implementation pattern, so I've made an example here 
 - https://github.com/srid/haskell-flake/pull/96
 
I think the main benefit is that you get the module syntax and module composition for free. It will be easy to define merge logic, which would be annoying to write with a custom type.

I think adding a `freeformType` to the package settings module would be a good idea. My example didn't include that, but it would allow undeclared parameters to be set, so that users won't be held back when they need to set something that's not explicitly added in haskell-flake yet.

---

Tried with freeform already and it was working using `evalModules`, but just curious to understand the advantage of one over the other. Thanks for the explanation and the PR.

---

> Also freeformType doesn't really support anything other than an attrset at its root. It would never get to see the module function.
But in the above shared example using `oneOf`, using functionTo which is not really a attrset at the root level but this is working with freeformType.

---

> [..] it would allow undeclared parameters to be set, so that users won't be held back when they need to set something that's not explicitly added in haskell-flake yet.

I actually think that using explicit options (as in #96), rather than a freeform attrset, is beneficial in that regard.

---

... unless we get as useful/simpler error messages when using freeformType (I'm not sure if we can).